src/Compilers/CSharp/Portable/Parser/Lexer_StringLiteral.cs

```diff
@@ -328,144 +328,580 @@ private void TrySetUnrecoverableError(SyntaxDiagnosticInfo error)
+                    // verbatim string beginning already.
+
+                    kind = InterpolatedStringKind.Verbatim;
+                    startingDollarSignCount = 1;
+                    startingQuoteCount = 1;
+                    window.AdvanceChar(3);
+                    return true;
                 }
-                else
+
+                if ((window.PeekChar(0), window.PeekChar(1), window.PeekChar(2), window.PeekChar(3)) is
+                        ('$', '"', not '"', _) or ('$', '"', '"', not '"'))
                 {
-                    Debug.Assert(_lexer.TextWindow.PeekChar() == '$');
-                    _lexer.TextWindow.AdvanceChar();
+                    // $"...
+                    // $""
+                    // not $"""
+                    kind = InterpolatedStringKind.Normal;
+                    startingDollarSignCount = 1;
+                    startingQuoteCount = 1;
+                    window.AdvanceChar(2);
+                    return true;
                 }
 
-                Debug.Assert(_lexer.TextWindow.PeekChar() == '"');
-                _lexer.TextWindow.AdvanceChar();
+                // From this point we have a raw literal of some sort.
+                var prefixAtCount = _lexer.ConsumeAtSignSequence();
+                startingDollarSignCount = _lexer.ConsumeDollarSignSequence();
+                var suffixAtCount = _lexer.ConsumeAtSignSequence();
+                startingQuoteCount = _lexer.ConsumeQuoteSequence();
+
+                var totalAtCount = prefixAtCount + suffixAtCount;
 
-                openQuoteRange = new Range(start, _lexer.TextWindow.Position);
+                // We should only have gotten here if we had at least two characters that made us think we had an interpolated string.
+                Debug.Assert(totalAtCount + startingDollarSignCount + startingQuoteCount >= 2);
+
+                // Multiple @-signs, and @-signs with raw literals are always illegal.  Detect these and give a
+                // reasonable error message.  Continue on if we can.
+                if (totalAtCount > 0)
+                {
+                    if (startingDollarSignCount == 0 && startingQuoteCount == 0)
+                    {
+                        // just multiple @'s in a row.  Give a general message about how @ signs work. We cannot
+                        // continue on as we have no quotes, and thus can't even find where the string starts or ends.
+                        TrySetError(_lexer.MakeError(start, width: 1, ErrorCode.ERR_ExpectedVerbatimLiteral));
+                        kind = InterpolatedStringKind.SingleLineRaw;
+                        return false;
+                    }
+
+                    // had at least a dollar sign or a quote.  pretend we have at least one dollar sign 
+                    // so we can at least try to consume the content.
+                    if (startingDollarSignCount == 0)
+                        startingDollarSignCount = 1;
```

Comment:
Why infer that we have an interpolated string as opposed to a non-interpolated string in this case (no `$` in code)? 
If we `startingDollarSignCount` as `0` then we could consume the content as content with no holes (braces interpreted as content).

Reply:
you're right!

Reply:
This has changed now.  we only ever get into interpolation parsign if we have at least one $.